3p/iframe-transport-client.js

```diff
@@ -33,61 +33,131 @@ export class IframeTransportClient {
         MessageType.SEND_IFRAME_TRANSPORT_EVENTS,
         MessageType.IFRAME_TRANSPORT_EVENTS,
         eventData => {
           const events =
-              /**
-               * @type
-               * {!Array<../src/3p-frame-messaging.IframeTransportEvent>}
-               */
-              (eventData['events']);
-          user().assert(events,
+          /**
+           * @type
+           *   {!Array<../src/3p-frame-messaging.IframeTransportEvent>}
+           */
+          (eventData['events']);
+          dev().assert(events,
               'Received malformed events list in ' + this.win_.location.href);
           dev().assert(events.length,
               'Received empty events list in ' + this.win_.location.href);
-          user().assert(this.listener_,
-              'Must call onAnalyticsEvent in ' + this.win_.location.href);
           events.forEach(event => {
             try {
-              this.listener_ &&
-                  this.listener_(event.message, event.transportId);
+              dev().assert(event.creativeId,
+                  'Received malformed event in ' + this.win_.location.href);
+              this.contextFor_(event.creativeId).dispatch(event.message);
             } catch (e) {
               user().error(TAG_,
                   'Exception in callback passed to onAnalyticsEvent: ' +
-                  e.message);
+                  e);
             }
           });
         });
   }
 
   /**
+   * Retrieves/creates a context object to pass events pertaining to a
+   * particular creative.
+   * @param {string} creativeId The ID of the creative
+   * @returns {!IframeTransportContext}
+   * @private
+   */
+  contextFor_(creativeId) {
+    return this.creativeIdToContext_[creativeId] ||
+        (this.creativeIdToContext_[creativeId] =
+            new IframeTransportContext(this.win_, this.iframeMessagingClient_,
+            creativeId, this.vendor_));
+  }
+
+  /**
+   * Gets the IframeMessagingClient.
+   * @returns {!IframeMessagingClient}
+   * @VisibleForTesting
+   */
+  getIframeMessagingClient() {
+    return this.iframeMessagingClient_;
+  }
+}
+
+/**
+ * A context object to be passed along with event data.
+ */
+class IframeTransportContext {
+  /**
+   * @param {!Window} win
+   * @param {!IframeMessagingClient} iframeMessagingClient
+   * @param {string} creativeId The ID of the creative that the event
+   *     pertains to.
+   * @param {string} vendor The 3p vendor name
+   */
+  constructor(win, iframeMessagingClient, creativeId, vendor) {
+    /** @private {!IframeMessagingClient} */
+    this.iframeMessagingClient_ = iframeMessagingClient;
+
+    /** @private @const {!Object} */
+    this.baseMessage_ = {creativeId, vendor};
+
+    /** @private {?function(string)} */
+    this.listener_ = null;
+
+    user().assert(win['onNewAmpAnalyticsInstance'],
+        'Must implement onNewAmpAnalyticsInstance in ' +
+        win.location.href);
+    win['onNewAmpAnalyticsInstance'](this);
```

Comment:
We're ok with failing on all future messages even if onNewAmpAnalyticsInstance didn't exist for the first?  Should we verify its a function?

Reply:
Added code to verify it's a function.
I'm fine with the failure you mentioned (at least for now - could always address it in a future PR) since per the example, customers are instructed to make sure that onNewAmpAnalyticsInstance exists before loading this library.